numpy/linalg/_linalg.py

```diff
@@ -2068,18 +2069,23 @@ def matrix_rank(A, tol=None, hermitian=False, *, rtol=None):
     >>> matrix_rank(np.zeros((4,)))
     0
     """
+    if rtol is not None and tol is not None:
+        raise ValueError("`tol` and `rtol` can't be both set.")
+
     A = asarray(A)
     if A.ndim < 2:
         return int(not all(A == 0))
     S = svd(A, compute_uv=False, hermitian=hermitian)
-    if rtol is not None and tol is not None:
-        raise ValueError("`tol` and `rtol` can't be both set.")
+
     if rtol is None:
         rtol = max(A.shape[-2:]) * finfo(S.dtype).eps
+    rtol = broadcast_to(rtol, A.shape[:-2])[..., newaxis]
```

Comment:
This should be inside the `if tol is None` branch. Indeed, the `if rtol is None` statement should be moved there - no point calculating `rtol` if it is not going to be used anyway.

Reply:
I changed it to `rtol = asarray(rtol)[..., newaxis]` as we still want to call `asarray` (similarly `asarray(tol)` later) to accept list inputs, like: `rtol = [0.2, 0.3, 0.1]`.